src/morx.rs

```diff
@@ -0,0 +1,1825 @@
+                }
+                //end of PERFORM_ACTION
+
+                if entry_flags & DONT_ADVANCE == 0 {
+                    break 'glyph; //exit the loop 'glyph unless entry_flags says DONT_ADVANCE
+                } else {
+                    self.component_stack.clear(); //if the entry_flags does say DONT_ADVANCE, then keep looping with the same glyph.
+                                                  //clear the stack
+                }
+            }
+            //end of loop 'glyph
+
+            i += 1; //advance to the next glyph
+        }
+        //end of loop 'glyphs
+
+        Ok(())
+    }
+}
+
+//------------------------------------ NonContextual Lookup ---------------------------------------------------------------
+//This function looks up and returns the noncontexutal substitute of glyph.
+//It returns 0xFFFF for a glyph index out of bounds of the lookup value array indices.
+pub fn noncontextual_lookup(glyph: u16, lookup_table: &ClassLookupTable) -> u16 {
+    match lookup(glyph, lookup_table) {
+        None => {
+            return 0xFFFF;
+        }
+        Some(val) => {
+            return val;
+        }
+    }
+}
+
+//------------------------------------ NonContextual Substiution ----------------------------------------------------------
+pub fn noncontextual_substitution(
+    glyphs: &mut Vec<RawGlyph<()>>,
+    noncontextual_subtable: &NonContextualSubtable,
+) -> Result<(), ParseError> {
+    let mut glyph: u16;
+    let mut subst: u16;
+    for i in 0..glyphs.len() {
+        glyph = glyphs[i].glyph_index;
+
+        subst = noncontextual_lookup(glyph, &noncontextual_subtable.lookup_table);
+
+        if (subst != 0xFFFF) && (subst != glyph) {
+            glyphs[i].glyph_index = subst;
+            glyphs[i].glyph_origin = GlyphOrigin::Direct;
+        }
+    }
+    Ok(())
+}
+
+//------------------------------------ Apply ligatures to an array of glyphs ----------------------------------------------
+pub fn apply(
+    morx_table: &MorxTable,
+    glyphs: &mut Vec<RawGlyph<()>>,
+    features: &Features,
+) -> Result<(), ParseError> {
+    for chain in morx_table.morx_chains.iter() {
+        let subfeatureflags: u32 = subfeatureflags(chain, features)?;
+        for subtable in chain.subtables.iter() {
+            if subfeatureflags & subtable.subtable_header.sub_feature_flags != 0 {
+                match subtable.subtable_header.coverage & 0xFF {
+                    1 => {
+                        let mut contextual_subst: ContextualSubstitution<'_> =
+                            ContextualSubstitution::new(glyphs);
+
+                        if let SubtableType::Contextual {
+                            contextual_subtable,
+                        } = &subtable.subtable_body
+                        {
+                            contextual_subst.next_state = 0;
+                            contextual_subst.process_glyphs(contextual_subtable)?;
+                        }
```

Comment:
Should there be `else` branches on these to return an error if the sub-table type is not the expected type?